File: kotlin/0207-course-schedule.kt
class Solution {
    fun canFinish(numCourses: Int, prerequisites: Array<IntArray>): Boolean {
      
      if (prerequisites.isEmpty()) return true
      
      val nodes = mutableMapOf<Int, MutableList<Int>>()
      prerequisites.forEach { 
        val (v, w) = it
        nodes.getOrPut(v) { mutableListOf<Int>() } += w  
      }
      
      val visited = mutableSetOf<Int>()
      val visitedInCurrentStack = mutableSetOf<Int>()
      var hasCycle = false
      
      fun traverse(startNode: Int) {
        
        visited.add(startNode)
        visitedInCurrentStack.add(startNode)
        
        for (v in nodes[startNode].orEmpty()) {
          if (!visited.contains(v)) {
            traverse(v)
          } else if (visitedInCurrentStack.contains(v)) {
            hasCycle = true
            return
          }
        }
        
        visitedInCurrentStack.remove(startNode)
      }
      
      for (key in nodes.keys) {
        traverse(key)  
      }
      
      return !hasCycle
    }
}

File: rust/0207-course-schedule.rs
impl Solution {
    pub fn can_finish(num_courses: i32, prerequisites: Vec<Vec<i32>>) -> bool {
        let mut graph = vec![vec![]; num_courses as usize];
        let mut visited = vec![0; num_courses as usize];

        for edge in prerequisites {
            graph[edge[0] as usize].push(edge[1] as usize);
        }

        for i in 0..num_courses as usize {
            if !Self::dfs(i, &graph, &mut visited) {
                return false;
            }
        }

        true
    }

    fn dfs(i: usize, graph: &Vec<Vec<usize>>, visited: &mut Vec<i32>) -> bool {
        if visited[i] == 1 {
            return false;
        }

        if visited[i] == -1 {
            return true;
        }

        visited[i] = 1;

        for j in graph[i].iter() {
            if !Self::dfs(*j, graph, visited) {
                return false;
            }
        }

        visited[i] = -1;

        true
    }
}
